lib/nvme/nvme_pcie.c

```diff
@@ -1841,7 +1841,24 @@ nvme_pcie_qpair_check_timeout(struct spdk_nvme_qpair *qpair)
 		 * Request has timed out. This could be i/o or admin request.
 		 * Call the registered timeout function for user to take action.
 		 */
-		ctrlr->timeout_cb_fn(ctrlr->timeout_cb_arg, ctrlr, qpair, tr->cid);
+
```

Comment:
This implementation is designed for a single I/O timeout value on a channel basis. It is for error recovery and therefore designed for minimal performance impact. We are simply allowing the user to try an abort instead of a reset.

It would be a great addition to allow a user to send an abort for any specific command for any reason. This would require a one-time traversal of the tracker queue to match the user's context. But, this is orthogonal to what we are trying to do here.

Another idea would be to allow the user to specify timeout values on a per command basis. I'm not sure how useful this is, but I do know our testing showed that traversing the tracker queue, even periodically, has measurable impact on latency.

The great thing about the callback is that the user can turn it off. With the ability to abort specific commands (which we should do), the user is free to be as fine-grained as they want.